src/emulator/module/include/module/read_arg.h

```diff
@@ -38,18 +38,27 @@ std::enable_if_t<sizeof(T) == 8, T> read_from_gpr(CPUState &cpu, const ArgLayout
     return static_cast<T>(both);
 }
 
+// Read float value from register.
+template <typename T>
+T read_from_fp(CPUState &cpu, const ArgLayout &arg) {
+    const float reg = read_float_reg(cpu, arg.offset);
+    return static_cast<T>(reg);
+}
+
 // Read variable from register or stack, as specified by arg layout.
 template <typename T>
 T read(CPUState &cpu, const ArgLayout &arg, const MemState &mem) {
     switch (arg.location) {
-        case ArgLocation::gpr:
-            return read_from_gpr<T>(cpu, arg);
-        case ArgLocation::stack:
-        {
-            const Address sp = read_sp(cpu);
-            const Address address_on_stack = static_cast<Address>(sp + arg.offset);
-            return *Ptr<T>(address_on_stack).get(mem);
-        }
+    case ArgLocation::gpr:
+        return read_from_gpr<T>(cpu, arg);
+    case ArgLocation::stack: {
+        const Address sp = read_sp(cpu);
+        const Address address_on_stack = static_cast<Address>(sp + arg.offset);
+        return *Ptr<T>(address_on_stack).get(mem);
+    }
+    case ArgLocation::fp:
+        if constexpr (std::is_same_v<T, float>)
+            return read_from_fp<T>(cpu, arg);
```

Comment:
This `if` should be redundant, as `arg.location` should always be `ArgLocation::fp` for `float`s. If that's not true, we're going to be returning default values for a bunch of non-float args that somehow get assigned to a floating point register. (`ArgLocation::fp`)

Reply:
Never mind, explained in Discord.